### DISCUSSION TITLE
Upload a File to SPO Doc Library in Web API Using an All In Memory Approach?

### DISCUSSION THREAD

**Discussion by RobertWoehr** [pnp/pnpcore#1393] [Q&A] [answered]
**Situational Background**

I work at a company that has a farm server side process that uploads files daily to a SharePoint 2013 document library.  The on-prem site has been migrated to SPO & due to firewall restrictions, the process will NOT be able to directly upload files to a SPO document library.  We have an internal to external gateway API that the process can utilize in order access an Azure Web API which is where I am able to make the upload file call utilizing PnP.Core in C#.

I have a controller action that reads the file from the request body & successfully uploads the file to the library, but I'm questioning if my approach could be improved?

In a nutshell, the request body is read into a MemoryStream then converted to a byte array which gets written to a file.  Next the file is read into a FileStream which PnP.Core uses as an arg in order to upload the file to the library.

The big question for this approach is:  Can I avoid writing & reading the file in the Web API before uploading & do it all in memory utilizing some sort of Stream object approach?

**Here's my code snippets**

`var boundary = HeaderUtilities.RemoveQuotes(
    MediaTypeHeaderValue.Parse(Request.ContentType).Boundary
).Value;

var reader = new MultipartReader(boundary, Request.Body);

var section = await reader.ReadNextSectionAsync();

// Use the PnP Context factory to get a PnPContext for the given configuration
using (var ctx = await pnpContextFactory.CreateAsync(sc))
{
    // Get a reference to a document library
    var folder = await ctx.Web.GetFolderByServerRelativeUrlAsync(ln);

    string filePath = Path.GetFullPath(Path.Combine(Environment.CurrentDirectory, "UploadedStreamedFiles"));

    if (await CreateFile(reader, section, filePath))
    {
        using (FileStream fs = System.IO.File.OpenRead(Path.Combine(filePath, fileName)))
        {
            IFile addedFile = await folder.Files.AddAsync(fn, fs, true);
        }
        System.IO.File.Delete(Path.Combine(filePath, fileName));

        return returnValue;
    }
    else
    {
        return "Could not create file.";
    }
}`

**Parse file from body & write it**

`private async Task<bool> CreateFile(MultipartReader reader, MultipartSection? section, string filePath)
{
    if (section != null)
    {
        var hasContentDispositionHeader = ContentDispositionHeaderValue.TryParse(
            section.ContentDisposition, out var contentDisposition
        );

        if (hasContentDispositionHeader)
        {
            if (contentDisposition.DispositionType.Equals("form-data") &&
            (!string.IsNullOrEmpty(contentDisposition.FileName.Value) ||
            !string.IsNullOrEmpty(contentDisposition.FileNameStar.Value)))
            {
                if (!Directory.Exists(filePath))
                {
                    Directory.CreateDirectory(filePath);
                }
                byte[] fileArray;
                using (var memoryStream = new MemoryStream())
                {
                    await section.Body.CopyToAsync(memoryStream);
                    fileArray = memoryStream.ToArray();
                }
                using (var fileStream = System.IO.File.Create(Path.Combine(filePath, contentDisposition.FileName.Value)))
                {
                    fileStream.Write(fileArray, 0, fileArray.Length);
                    return true;
                }
            }
            else
            {
                return false;
            }
        }
        else
        {
            return false;
        }
    }
    else
    {
        return false;
    }
}`

Any advice would be appeciated!!!

Thanks!

**Comment by jansenbe** [comment]
@RobertWoehr : the AddAsync method on the `IFilesCollection` takes a `Stream` as input (see https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#adding-files-uploading and https://pnp.github.io/pnpcore/api/PnP.Core.Model.SharePoint.IFileCollection.html#PnP_Core_Model_SharePoint_IFileCollection_AddAsync_System_String_Stream_System_Boolean_)...so why don't you set the Stream.Position to 0 and reuse the stream?

**Reply by RobertWoehr** [reply, marked as answer]
@jansenbe Wow!  You're my Hero!!!  And such a simple answer!  I tested your suggestion out on .xlsx, .docx, & .pdf and it seems to be working as expected!  Thank you for such a fantastic & timely response!
